RFC-001-libp2p.adoc

```diff
@@ -0,0 +1,361 @@
+The listening party has to have a port open the dialling party can connect to.
+If one party is behind a NAT, the only way to establish a network connection is for this party to dial to the other one.
+Establishing such a connection is expensive.
+We have to wait for the TCP handshake to complete and if there is transport encryption in place, keys might need to be exchanged before the first actual payload can be sent to the other party.
+Due to this cost and network limitations like NAT, it is good to reuse a connection between two parties once we have established one.
+This is where multiplexing comes into play.
+If both parties want to be able to send and receive messages at any point in time, those messages need to carry metadata so that the receiving end can correctly identify and associate messages.
+
+If all you have are plain network sockets, such a multiplexing strategy needs to be backed into your application protocol, which is a none trivial increase in complexity.
+In fact, a former version of the COMIT messaging protocol (BAM!) had support for such functionality because we were not yet building on top of libp2p back then.
+
+==== Multiplexing with libp2p substreams
+
+Libp2p has the concept of substreams.
+A substream is a virtual socket for a specific protocol between two nodes.
+Opening a new substream is cheap compared to an actual network connection.
+It is also independent of the underlying physical network topology, i.e. it does not matter which party opened the actual physical network connection:
+Every party can open a new substream at any point in time.
+
+=== Design
+
+This paragraph describes the over-arching design of the COMIT messaging protocol.
+It builds heavily upon the https://github.com/libp2p/specs/tree/master/connections#definitions[terminology] that is used within the libp2p spec.
+
+To send a new message to another node, a COMIT node MUST open a new substream.
+Within a single substream, there is always just one _active_ party.
+Only the _active_ is allowed to send a message.
+The dialer of the substream is initially the _active_ party.
+Hence, the node opening the substream is the first one to send a message.
+
+==== Message format
+
+All messages within the COMIT protocol are JSON objects.
+They MUST be serialized to a single line and are delimited by the newline character `\n`.
+The messages MUST use UTF-8 encoding.
+
+TODO: include pretty printing note
+
+==== Dealing with errors
+
+Nodes MAY immediately close the substream if they encounter an error.
+This includes:
+
+- deserialization errors: i.e. the other party sends malformed JSON
+- protocol errors: the other party does not conform to the messaging protocol
+
+Implementations SHOULD consider it a hard error a remote node preemptively closes a substream.
+Very likely, the substream got closed because an unrecoverable error (like in the list above) and hence, implementations SHOULD NOT retry the communication because it is unlikely to succeed.
```

Comment:
```suggestion
Very likely, the substream got closed because of an unrecoverable error (like in the list above) and hence, implementations SHOULD NOT retry the communication because it is unlikely to succeed.
```